File: ios/Packages/FatwaBotKit/Sources/ContentKit/ContentService.swift
import CoreKit
import Foundation
import OSLog
import NetworkingKit

/// Content sync (docs/features/content-pipeline.md). Mirrors ConfigService's
/// offline-first shape: `current(locale:)` always resolves synchronously
/// (cache → bundled seed → nil), `refresh` fetches deltas and is silent on
/// failure — content stays fully usable with zero connectivity.
public actor ContentService {
    private let logger = Logger(subsystem: "com.fatwabot.app", category: "content-sync")
    private let store: ContentFileStore
    private let client: APIClientProtocol

    private var cachedAzkar: [String: AzkarCollection] = [:]
    private var cachedDuas: [String: DuaCollection] = [:]
    private var cachedHadithSummaries: [String: [HadithCollectionSummary]] = [:]
    private var cachedHadithDetail: [String: HadithCollectionDetail] = [:] // key: "\(slug)_\(locale)"
    private var cachedWird: [String: WirdTemplatesCollection] = [:]

    public init(store: ContentFileStore, client: APIClientProtocol) {
        self.store = store
        self.client = client
    }

    // MARK: - Synchronous reads (cache → bundled seed)

    public func azkar(locale: String) -> AzkarCollection? {
        if let cached = cachedAzkar[locale] { return cached }
        let loaded = store.load(AzkarCollection.self, key: "azkar.\(locale)")
        cachedAzkar[locale] = loaded
        return loaded
    }

    public func duas(locale: String) -> DuaCollection? {
        if let cached = cachedDuas[locale] { return cached }
        let loaded = store.load(DuaCollection.self, key: "duas.\(locale)")
        cachedDuas[locale] = loaded
        return loaded
    }

    public func hadithCollections(locale: String) -> [HadithCollectionSummary] {
        if let cached = cachedHadithSummaries[locale] { return cached }
        let loaded = store.load(HadithCollectionsResponse.self, key: "hadith-collections.\(locale)")?.collections ?? []
        cachedHadithSummaries[locale] = loaded
        return loaded
    }

    public func hadithDetail(slug: String, locale: String) -> HadithCollectionDetail? {
        let key = "\(slug)_\(locale)"
        if let cached = cachedHadithDetail[key] { return cached }
        let loaded = store.load(HadithCollectionDetail.self, key: "hadith-\(slug).\(locale)")
        cachedHadithDetail[key] = loaded
        return loaded
    }

    public func wirdTemplates(locale: String) -> WirdTemplatesCollection? {
        if let cached = cachedWird[locale] { return cached }
        let loaded = store.load(WirdTemplatesCollection.self, key: "wird-templates.\(locale)")
        cachedWird[locale] = loaded
        return loaded
    }

    // MARK: - Refresh (per-collection independent failure)

    /// Why a refresh did not change anything — the distinction `try?` erased.
    ///
    /// Every `refresh*` used to return `false` for both "the server had nothing
    /// new" and "the request failed", which made a broken sync indistinguishable
    /// from a healthy one. That is how an entire content release went invisible:
    /// the app kept serving a stale cache and nothing, anywhere, said so.
    ///
    /// Failures are still non-fatal — the app is offline-first and the cached
    /// copy is a valid answer — but they are now recorded and logged rather than
    /// swallowed.
    public enum RefreshOutcome: Equatable, Sendable {
        case updated
        case unchanged
        case failed(String)

        public var didUpdate: Bool { self == .updated }
        public var isFailure: Bool { if case .failed = self { return true }; return false }
    }

    /// Last failure per content key, for diagnostics. Cleared on success.
    public private(set) var lastFailures: [String: String] = [:]

    private func record(_ key: String, _ outcome: RefreshOutcome) -> RefreshOutcome {
        if case let .failed(message) = outcome {
            lastFailures[key] = message
            logger.error("content refresh failed for \(key, privacy: .public): \(message, privacy: .public)")
        } else {
            lastFailures.removeValue(forKey: key)
        }
        return outcome
    }

    /// Runs one fetch, converting a thrown error into an observable outcome.
    private func refresh<T>(
        key: String,
        endpoint: Endpoint<T>,
        isUnchanged: (T) -> Bool,
        apply: (T) -> Void
    ) async -> RefreshOutcome {
        do {
            let fresh = try await client.get(endpoint)
            if isUnchanged(fresh) { return record(key, .unchanged) }
            apply(fresh)
            return record(key, .updated)
        } catch {
            return record(key, .failed(String(describing: error)))
        }
    }

    @discardableResult
    public func refreshAzkar(locale: String) async -> RefreshOutcome {
        await refresh(
            key: "azkar.\(locale)",
            endpoint: ContentEndpoints.azkar(sinceVersion: azkar(locale: locale)?.version),
            isUnchanged: { $0 == self.cachedAzkar[locale] },
            apply: { fresh in
                self.cachedAzkar[locale] = fresh
                self.store.save(fresh, key: "azkar.\(locale)")
            }
        )
    }

    @discardableResult
    public func refreshDuas(locale: String) async -> RefreshOutcome {
        await refresh(
            key: "duas.\(locale)",
            endpoint: ContentEndpoints.duas(sinceVersion: duas(locale: locale)?.version),
            isUnchanged: { $0 == self.cachedDuas[locale] },
            apply: { fresh in
                self.cachedDuas[locale] = fresh
                self.store.save(fresh, key: "duas.\(locale)")
            }
        )
    }

    @discardableResult
    public func refreshHadithCollections(locale: String) async -> RefreshOutcome {
        await refresh(
            key: "hadith-collections.\(locale)",
            endpoint: ContentEndpoints.hadithCollections,
            isUnchanged: { $0.collections == self.cachedHadithSummaries[locale] },
            apply: { fresh in
                self.cachedHadithSummaries[locale] = fresh.collections
                self.store.save(fresh, key: "hadith-collections.\(locale)")
            }
        )
    }

    @discardableResult
    public func refreshHadithDetail(slug: String, locale: String) async -> RefreshOutcome {
        let cacheKey = "\(slug)_\(locale)"
        return await refresh(
            key: "hadith-\(slug).\(locale)",
            endpoint: ContentEndpoints.hadithDetail(
                slug: slug,
                sinceVersion: hadithDetail(slug: slug, locale: locale)?.version
            ),
            isUnchanged: { $0 == self.cachedHadithDetail[cacheKey] },
            apply: { fresh in
                self.cachedHadithDetail[cacheKey] = fresh
                self.store.save(fresh, key: "hadith-\(slug).\(locale)")
            }
        )
    }

    @discardableResult
    public func refreshWirdTemplates(locale: String) async -> RefreshOutcome {
        await refresh(
            key: "wird-templates.\(locale)",
            endpoint: ContentEndpoints.wirdTemplates(sinceVersion: wirdTemplates(locale: locale)?.version),
            isUnchanged: { $0 == self.cachedWird[locale] },
            apply: { fresh in
                self.cachedWird[locale] = fresh
                self.store.save(fresh, key: "wird-templates.\(locale)")
            }
        )
    }
}

File: android/core/content/src/main/kotlin/com/fatwabot/core/content/ContentService.kt
package com.fatwabot.core.content

import com.fatwabot.core.network.ApiClientProtocol
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.serialization.json.Json

/**
 * Content sync — mirror of iOS ContentKit.ContentService. `current` reads are
 * synchronous (cache -> bundled seed -> null); refresh is per-collection
 * independent and silent on failure, so content stays fully usable offline.
 */
class ContentService(
    private val store: ContentFileStore,
    private val client: ApiClientProtocol,
) {
    private val json = Json { ignoreUnknownKeys = true }
    private val mutex = Mutex()

    private val cachedAzkar = mutableMapOf<String, AzkarCollection>()
    private val cachedDuas = mutableMapOf<String, DuaCollection>()
    private val cachedHadithSummaries = mutableMapOf<String, List<HadithCollectionSummary>>()
    private val cachedHadithDetail = mutableMapOf<String, HadithCollectionDetail>() // key: "slug_locale"
    private val cachedWird = mutableMapOf<String, WirdTemplatesCollection>()

    fun azkar(locale: String): AzkarCollection? = cachedAzkar.getOrPut(locale) {
        store.load(AzkarCollection.serializer(), "azkar.$locale") ?: return null
    }

    fun duas(locale: String): DuaCollection? = cachedDuas.getOrPut(locale) {
        store.load(DuaCollection.serializer(), "duas.$locale") ?: return null
    }

    fun hadithCollections(locale: String): List<HadithCollectionSummary> = cachedHadithSummaries.getOrPut(locale) {
        store.load(HadithCollectionsResponse.serializer(), "hadith-collections.$locale")?.collections ?: emptyList()
    }

    fun hadithDetail(slug: String, locale: String): HadithCollectionDetail? {
        val key = "${slug}_$locale"
        return cachedHadithDetail.getOrPut(key) {
            store.load(HadithCollectionDetail.serializer(), "hadith-$slug.$locale") ?: return null
        }
    }

    fun wirdTemplates(locale: String): WirdTemplatesCollection? = cachedWird.getOrPut(locale) {
        store.load(WirdTemplatesCollection.serializer(), "wird-templates.$locale") ?: return null
    }

    suspend fun refreshAzkar(locale: String): Boolean = mutex.withLock {
        val since = azkar(locale)?.version
        val query = sinceQuery(since)
        val raw = runCatching { client.getRaw("v1/content/azkar", query) }.getOrNull() ?: return false
        val fresh = runCatching { json.decodeFromString(AzkarCollection.serializer(), raw) }.getOrNull() ?: return false
        if (fresh == cachedAzkar[locale]) return false
        cachedAzkar[locale] = fresh
        store.save(AzkarCollection.serializer(), fresh, "azkar.$locale")
        true
    }

    suspend fun refreshDuas(locale: String): Boolean = mutex.withLock {
        val since = duas(locale)?.version
        val raw = runCatching { client.getRaw("v1/content/duas", sinceQuery(since)) }.getOrNull() ?: return false
        val fresh = runCatching { json.decodeFromString(DuaCollection.serializer(), raw) }.getOrNull() ?: return false
        if (fresh == cachedDuas[locale]) return false
        cachedDuas[locale] = fresh
        store.save(DuaCollection.serializer(), fresh, "duas.$locale")
        true
    }

    suspend fun refreshHadithCollections(locale: String): Boolean = mutex.withLock {
        val raw = runCatching { client.getRaw("v1/content/hadith-collections") }.getOrNull() ?: return false
        val fresh = runCatching {
            json.decodeFromString(HadithCollectionsResponse.serializer(), raw)
        }.getOrNull() ?: return false
        if (fresh.collections == cachedHadithSummaries[locale]) return false
        cachedHadithSummaries[locale] = fresh.collections
        store.save(HadithCollectionsResponse.serializer(), fresh, "hadith-collections.$locale")
        true
    }

    suspend fun refreshHadithDetail(slug: String, locale: String): Boolean = mutex.withLock {
        val key = "${slug}_$locale"
        val since = hadithDetail(slug, locale)?.version
        val raw = runCatching {
            client.getRaw("v1/content/hadith-collections/$slug", sinceQuery(since))
        }.getOrNull() ?: return false
        val fresh = runCatching {
            json.decodeFromString(HadithCollectionDetail.serializer(), raw)
        }.getOrNull() ?: return false
        if (fresh == cachedHadithDetail[key]) return false
        cachedHadithDetail[key] = fresh
        store.save(HadithCollectionDetail.serializer(), fresh, "hadith-$slug.$locale")
        true
    }

    suspend fun refreshWirdTemplates(locale: String): Boolean = mutex.withLock {
        val since = wirdTemplates(locale)?.version
        val raw = runCatching { client.getRaw("v1/content/wird-templates", sinceQuery(since)) }.getOrNull() ?: return false
        val fresh = runCatching {
            json.decodeFromString(WirdTemplatesCollection.serializer(), raw)
        }.getOrNull() ?: return false
        if (fresh == cachedWird[locale]) return false
        cachedWird[locale] = fresh
        store.save(WirdTemplatesCollection.serializer(), fresh, "wird-templates.$locale")
        true
    }

    private fun sinceQuery(version: Int?): Map<String, String> =
        version?.let { mapOf("since_version" to it.toString()) } ?: emptyMap()
}
